colexec: clean up memory infra on panic in NewColOperator

yuzefovich opened the pull request:
It is possible that an unexpected error occurs during NewColOperator
call after we have created some memory monitoring infrastructure. In
order to correctly clean up the infra in such scenario we will now have
a separate panic catcher that closes the infra within NewColOperator.
We will also now clean up the infra when an error occurs.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/44564)
<!-- Reviewable:end -->

asubiotto reviewed (commented):
I think it makes more sense to do this panic catching and cleanup inside `NewColOperator` (and any cleanup as well), seems like the current abstraction is a bit leaky if the caller is expected to clean up state it doesn't know about.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44564#-:-M-AVOe26_mY8wLxjDds:bp4dwpk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)



<!-- Sent from Reviewable.io -->

yuzefovich committed:
colexec: clean up memory infra on panic in NewColOperator

It is possible that an unexpected error occurs during NewColOperator
call after we have created some memory monitoring infrastructure. In
order to correctly clean up the infra in such scenario we will now have
a separate panic catcher that closes the infra within NewColOperator.
We will also now clean up the infra when an error occurs.

Release note: None

yuzefovich reviewed (commented):
Good point, done.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44564#-:-M-B420p-A4JQZUW5km5:bfbq99i)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44564)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/execplan.go, line 362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44564#-M-B5rzs3frlajVixrrB:-M-B5rzs3frlajVixrrC:b-8k4oan) ([raw file](https://github.com/cockroachdb/cockroach/blob/1350f2f6ddc5bd6b9b061140effa0ea9c1974ddb/pkg/sql/colexec/execplan.go#L362)):*
> ```Go
> ) (result NewColOperatorResult, err error) {
> 	// Make sure that we clean up memory monitoring infrastructure in case of a
> 	// panic.
> ```

It'll also be good to do this on error so that the caller doesn't have to


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44564)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/execplan.go, line 362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44564#-M-B5rzs3frlajVixrrB:-M-B6_BU5X7Sy6I8KOVx:bgnf5wf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1350f2f6ddc5bd6b9b061140effa0ea9c1974ddb/pkg/sql/colexec/execplan.go#L362)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

It'll also be good to do this on error so that the caller doesn't have to
</blockquote></details>

Hm, I don't really like this idea but not sure why. I feel like the caller will already have the code to clean up the infra (for the non-error case), so it might as well clean up the infra in the error case.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44564)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/execplan.go, line 362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44564#-M-B5rzs3frlajVixrrB:-M-B9A3o9RNv_xahZ1v4:bvqfh9q) ([raw file](https://github.com/cockroachdb/cockroach/blob/1350f2f6ddc5bd6b9b061140effa0ea9c1974ddb/pkg/sql/colexec/execplan.go#L362)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Hm, I don't really like this idea but not sure why. I feel like the caller will already have the code to clean up the infra (for the non-error case), so it might as well clean up the infra in the error case.
</blockquote></details>

I think that the caller should never have to worry about intermediate state that wasn't cleaned up on error when calling a function. Up to you, since we already have code that does that


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44564)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/execplan.go, line 362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44564#-M-B5rzs3frlajVixrrB:-M-BFn7i1A6DMnTVqyqp:bd1ko8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/1350f2f6ddc5bd6b9b061140effa0ea9c1974ddb/pkg/sql/colexec/execplan.go#L362)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I think that the caller should never have to worry about intermediate state that wasn't cleaned up on error when calling a function. Up to you, since we already have code that does that
</blockquote></details>

Alright, this sounds reasonable, done. I think this will work, right?


<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44564)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/execplan.go, line 362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44564#-M-B5rzs3frlajVixrrB:-M-BUVay6tx1JY_PzYpK:b-jktrq0) ([raw file](https://github.com/cockroachdb/cockroach/blob/1350f2f6ddc5bd6b9b061140effa0ea9c1974ddb/pkg/sql/colexec/execplan.go#L362)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Alright, this sounds reasonable, done. I think this will work, right?
</blockquote></details>

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, maybe we can remove the cleanup in the code that calls `setupVectorized`?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44564#-:-M-GJYOz-WqMQcHRZo_h:b-oduykh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/execplan.go, line 362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44564#-M-B5rzs3frlajVixrrB:-M-GJ9aJAC4XilYZ0D5V:b-9d9gut) ([raw file](https://github.com/cockroachdb/cockroach/blob/1350f2f6ddc5bd6b9b061140effa0ea9c1974ddb/pkg/sql/colexec/execplan.go#L362)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

:lgtm:, maybe we can remove the cleanup in the code that calls `setupVectorized`?
</blockquote></details>

Which code are you referring to? Maybe cleaning up in `SupportsVectorized`? In any case, I don't think we can remove any code because the caller is still responsible for clean up in non-error case, the only thing I can think of is that we could now add an `if err == nil` to it, but I don't think it'd help in any way.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44564)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/execplan.go, line 362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44564#-M-B5rzs3frlajVixrrB:-M-GNah30EfMjwtR6C4n:b-k0gbxp) ([raw file](https://github.com/cockroachdb/cockroach/blob/1350f2f6ddc5bd6b9b061140effa0ea9c1974ddb/pkg/sql/colexec/execplan.go#L362)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Which code are you referring to? Maybe cleaning up in `SupportsVectorized`? In any case, I don't think we can remove any code because the caller is still responsible for clean up in non-error case, the only thing I can think of is that we could now add an `if err == nil` to it, but I don't think it'd help in any way.
</blockquote></details>

I was thinking of `setupVectorized`, which still does stuff if `err != nil` (if `err == nil` the cleanup happens on `flow.Cleanup()`)


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Oh, I see. I think it should be possible.

bors r-

yuzefovich commented:
Actually, I take that back - we're creating memory infra for non-operators as well (hash router, synchronizers), and that happens outside of `NewColOperator`, so we still need to clean up all the infra even in non-error case.

bors r+
